lightstep/sdk/metric/aggregator/histogram/exponential.go

```diff
@@ -324,75 +308,39 @@ func (b *buckets) relocateBucket(dest, src int32) {
 	if dest == src {
 		return
 	}
-	switch counts := b.backing.(type) {
-	case []uint8:
-		tmp := counts[src]
-		counts[src] = 0
-		b.incrementBucket(dest, uint64(tmp))
-	case []uint16:
-		tmp := counts[src]
-		counts[src] = 0
-		b.incrementBucket(dest, uint64(tmp))
-	case []uint32:
-		tmp := counts[src]
-		counts[src] = 0
-		b.incrementBucket(dest, uint64(tmp))
-	case []uint64:
-		tmp := counts[src]
-		counts[src] = 0
-		b.incrementBucket(dest, uint64(tmp))
-	}
+
+	b.incrementBucket(dest, b.backing.moveCount(src))
 }
 
 // incrementBucket increments the backing array index by `incr`.
 func (b *buckets) incrementBucket(bucketIndex int32, incr uint64) {
 	for {
-		switch counts := b.backing.(type) {
-		case []uint8:
-			if uint64(counts[bucketIndex])+incr < 0x100 {
-				counts[bucketIndex] += uint8(incr)
-				return
-			}
-			tmp := make([]uint16, len(counts))
-			for i := range counts {
-				tmp[i] = uint16(counts[i])
-			}
-			b.backing = tmp
-			continue
-		case []uint16:
-			if uint64(counts[bucketIndex])+incr < 0x10000 {
-				counts[bucketIndex] += uint16(incr)
-				return
-			}
-			tmp := make([]uint32, len(counts))
-			for i := range counts {
-				tmp[i] = uint32(counts[i])
-			}
-			b.backing = tmp
-			continue
-		case []uint32:
-			if uint64(counts[bucketIndex])+incr < 0x100000000 {
-				counts[bucketIndex] += uint32(incr)
-				return
-			}
-			tmp := make([]uint64, len(counts))
-			for i := range counts {
-				tmp[i] = uint64(counts[i])
-			}
-			b.backing = tmp
-			continue
-		case []uint64:
-			counts[bucketIndex] += incr
+		if b.backing.tryIncrement(bucketIndex, incr) {
```

Comment:
I find this much more readily comprehensible now, thank you!

Reply:
Thanks! As you noticed, I made the histogram generic since the first draft was written so long before Go-1.18 landed.